libraries/AP_AccelCal/AccelCalibrator.cpp

```diff
@@ -0,0 +1,464 @@
+        clear();
+    }
+    if (_status != ACCEL_CAL_NOT_STARTED) {
+        return;
+    }
+
+    _conf_num_samples = num_samples;
+    _conf_sample_time = sample_time;
+    _conf_fit_type = fit_type;
+
+    _param_struct.offset = offset;
+    _param_struct.diag = diag;
+    _param_struct.offdiag = offdiag;
+
+    switch (_conf_fit_type) {
+        case ACCEL_CAL_AXIS_ALIGNED_ELLIPSOID:
+            if (_conf_num_samples < 6) {
+                set_status(ACCEL_CAL_FAILED);
+                return;
+            }
+            break;
+        case ACCEL_CAL_ELLIPSOID:
+            if (_conf_num_samples < 8) {
+                set_status(ACCEL_CAL_FAILED);
+                return;
+            }
+            break;
+    }
+
+    set_status(ACCEL_CAL_WAITING_FOR_ORIENTATION);
+}
+
+
+// set Accel calibrator status to make itself ready for future accel cals
+void AccelCalibrator::clear() {
+    set_status(ACCEL_CAL_NOT_STARTED);
+}
+
+// returns true if accel calibrator is running
+bool AccelCalibrator::running() {
+    return _status == ACCEL_CAL_WAITING_FOR_ORIENTATION || _status == ACCEL_CAL_COLLECTING_SAMPLE;
+}
+
+// set Accel calibrator to start collecting samples in the next cycle
+void AccelCalibrator::collect_sample() {
+    set_status(ACCEL_CAL_COLLECTING_SAMPLE);
+}
+
+// collect and avg sample to be passed onto LSQ estimator after all requisite orientations are done
+void AccelCalibrator::new_sample(Vector3f delta_velocity, float dt) {
+    if (_status != ACCEL_CAL_COLLECTING_SAMPLE) {
+        return;
+    }
+
+    if (_samples_collected >= _conf_num_samples) {
+        set_status(ACCEL_CAL_FAILED);
+        return;
+    }
+
+    _sample_buffer[_samples_collected].delta_velocity += delta_velocity;
+    _sample_buffer[_samples_collected].delta_time += dt;
+
+    _last_samp_frag_collected_ms = hal.scheduler->millis();
+
+    if (_sample_buffer[_samples_collected].delta_time > _conf_sample_time) {
+        Vector3f sample = _sample_buffer[_samples_collected].delta_velocity/_sample_buffer[_samples_collected].delta_time;
+        if (!accept_sample(sample)) {
+            set_status(ACCEL_CAL_FAILED);
+            return;
+        }
+
+        _samples_collected++;
+
+        if (_samples_collected >= _conf_num_samples) {
+            run_fit(MAX_ITERATIONS, _fitness);
+
+            if (_fitness < _conf_tolerance) {
+                set_status(ACCEL_CAL_SUCCESS);
+            } else {
+                set_status(ACCEL_CAL_FAILED);
+            }
+        } else {
+            set_status(ACCEL_CAL_WAITING_FOR_ORIENTATION);
+        }
+    }
+}
+
+// interface for LSq estimator to read sample buffer sent after conversion from delta velocity
+// to averaged acc over time
+bool AccelCalibrator::get_sample(uint8_t i, Vector3f& s) const {
+    if (i >= _samples_collected) {
+        return false;
+    }
+    s = _sample_buffer[i].delta_velocity / _sample_buffer[i].delta_time;
```

Comment:
if this has just stored an accel value you could have avoided the division and returned a const reference. I don't understand why storing this as a velocity and time is useful